### PR TITLE
operator: Custom configuration for LokiStack admin groups

### PR DESCRIPTION
**Pull request by aminesnow**
**What this PR does / why we need it**:
This PR introduces a new attribute `adminGroups` to the `LokiStack` CRD. This new attribute allows the user to define which groups to be considered by the operator as cluster admins.
The default cluster admin groups are: `system:cluster-admins`, `cluster-admin` and `dedicated-admin`.

**Special notes for your reviewer**:
\cc @periklis @xperimental 

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: operator/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ## Main
-
+- [9931](https://github.com/grafana/loki/pull/9931) **aminesnow**: Custom configuration for LokiStack admin groups
```
Comment (periklis) [inline operator/CHANGELOG.md:2]:
```suggestion

- [9931](https://github.com/grafana/loki/pull/9931) **aminesnow**: Custom configuration for LokiStack admin groups
```

File: operator/apis/loki/v1/lokistack_types.go [hunk 2]
```diff
@@ -838,6 +838,13 @@ type LokiStackSpec struct {
 	// +kubebuilder:validation:Optional
 	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Tenants Configuration"
 	Tenants *TenantsSpec `json:"tenants,omitempty"`
+
+	// AdminGroups defines the list of group to be considered cluster admins by the Loki Operator.
+	//
+	// +optional
+	// +kubebuilder:validation:Optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Admin Groups"
+	AdminGroups []string `json:"adminGroups,omitempty"`
```
Comment (periklis) [inline operator/apis/loki/v1/lokistack_types.go:847]: This field declares configuration that is not generally applicable. In fact is only applicable for tenant modes `openshift-logging` and `openshift-network`. Therefore I suggest to wrap this withing the tenants spec, e.g.:
```yaml
spec:
  tenants:
    mode: openshift-logging
    openshiftLogging:
      adminGroups:
      - a
      - b
      - c
```

**Review by periklis — CHANGES_REQUESTED** [review]

File: operator/internal/manifests/openshift/opa_openshift.go [hunk 3]
```diff
@@ -38,14 +38,17 @@ func newOPAOpenShiftContainer(mode lokiv1.ModeType, secretVolumeName, tlsDir, mi
...
 		fmt.Sprintf("--web.listen=:%d", GatewayOPAHTTPPort),
 		fmt.Sprintf("--web.internal.listen=:%d", GatewayOPAInternalPort),
 		fmt.Sprintf("--web.healthchecks.url=http://localhost:%d", GatewayOPAHTTPPort),
+		"--opa.skip-tenants=audit,infrastructure",
 		fmt.Sprintf("--opa.package=%s", opaDefaultPackage),
 	}
 
+	if len(adminGroups) > 0 {
+		args = append(args, fmt.Sprintf("--opa.admin-groups=%s", adminGroups))
+	}
```
Comment (periklis) [inline operator/internal/manifests/openshift/opa_openshift.go:50]: This is the solution when a user unsets the admin groups entirely, right?

File: operator/apis/loki/v1/lokistack_types.go [hunk 4]
```diff
@@ -244,6 +244,23 @@ type TenantsSpec struct {
...
+}
+
+// OpenshiftLoggingSpec defines the configuration specific to Openshift modes.
+type OpenshiftLoggingSpec struct {
+	// AdminGroups defines the list of group to be considered cluster admins by the Loki Operator.
+	//
+	// +optional
+	// +kubebuilder:validation:Optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Admin Groups"
+	AdminGroups []string `json:"adminGroups"`
```
[the remark this answers is not in the document]
Comment (aminesnow) [reply operator/apis/loki/v1/lokistack_types.go:263]: We can rename it to something like `openshiftConfig` if you think that would be less confusing.

File: operator/internal/manifests/openshift/opa_openshift.go (hunk 3 above)
Comment (aminesnow) [reply operator/internal/manifests/openshift/opa_openshift.go:50]: Yes, but it's only a part of the logic.

File: operator/apis/loki/v1/lokistack_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (xperimental) [reply operator/apis/loki/v1/lokistack_types.go:263]: I don't know if you two have already discussed this option: As we currently only have one "openshift option" which applies to both `openshift-logging` and `openshift-network` tenancy modes, what about renaming the spec to `OpenshiftTenantSpec` and, once we have options that are only valid for either `logging` or `network` add another level below `openshift`?

```suggestion

	// OpenshiftTenantSpec defines the configuration specific to Openshift modes.
	//
	// +optional
	// +kubebuilder:validation:Optional
	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Openshift"
	Openshift *OpenshiftTenantSpec `json:"openshift,omitempty"`
}

// OpenshiftTenantSpec defines the configuration specific to Openshift modes.
type OpenshiftTenantSpec struct {
	// AdminGroups defines a list of groups, whose members are considered to have admin-privileges by the Loki Operator.
	// Setting this to an empty array disables admin groups.
	//
	// By default the following groups are considered admin-groups:
	//
	//  - system:cluster-admins
	//  - cluster-admin
	//  - dedicated-admin
	//
	// +optional
	// +kubebuilder:validation:Optional
	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Admin Groups"
	AdminGroups []string `json:"adminGroups"`
```

(drive-by: added a bit more detail to the description of `AdminGroups`)

File: operator/internal/manifests/openshift/opa_openshift.go [hunk 5]
```diff
@@ -38,14 +38,17 @@ func newOPAOpenShiftContainer(mode lokiv1.ModeType, secretVolumeName, tlsDir, mi
 	uriScheme = corev1.URISchemeHTTP
 	args = []string{
 		"--log.level=warn",
-		"--opa.skip-tenants=audit,infrastructure",
-		"--opa.admin-groups=system:cluster-admins,cluster-admin,dedicated-admin",
 		fmt.Sprintf("--web.listen=:%d", GatewayOPAHTTPPort),
 		fmt.Sprintf("--web.internal.listen=:%d", GatewayOPAInternalPort),
 		fmt.Sprintf("--web.healthchecks.url=http://localhost:%d", GatewayOPAHTTPPort),
+		"--opa.skip-tenants=audit,infrastructure",
```
Comment (xperimental) [inline operator/internal/manifests/openshift/opa_openshift.go:44]: Do we want to keep the constant arguments at the top?

Comment (aminesnow) [reply operator/internal/manifests/openshift/opa_openshift.go:44]: I'm more interested in keeping in keeping arguments that are related together (opa, web.. etc)

File: operator/internal/manifests/gateway_tenants.go [hunk 6]
```diff
@@ -62,13 +64,18 @@ func ApplyGatewayDefaultOptions(opts *Options) error {
...
 		return nil // nothing to configure
 	case lokiv1.OpenshiftLogging, lokiv1.OpenshiftNetwork:
 		tlsDir := gatewayServerHTTPTLSDir()
-		return openshift.ConfigureGatewayDeployment(d, mode, tlsSecretVolume, tlsDir, minTLSVersion, ciphers, fg.HTTPEncryption)
+		adminGroupsStr := "system:cluster-admins,cluster-admin,dedicated-admin"
+		if adminGroups != nil {
+			adminGroupsStr = strings.Join(adminGroups, ",")
+		}
+
+		return openshift.ConfigureGatewayDeployment(d, mode, tlsSecretVolume, tlsDir, minTLSVersion, ciphers, fg.HTTPEncryption, adminGroupsStr)
```
[the remark this answers is not in the document]
Comment (aminesnow) [reply operator/internal/manifests/gateway_tenants.go:78]: Slices all the way down!

**Review by periklis — APPROVED** [review]

**Pull request merged**